deepchem/metalearning/maml.py

```diff
@@ -0,0 +1,220 @@
+    self.save_file = "%s/%s" % (self.model_dir, "model")
+    with self._graph.as_default():
+      # Create duplicate placeholders for meta-optimization.
+
+      learner.select_task()
+      self._meta_placeholders = {}
+      for p in learner.get_batch().keys():
+        name = 'meta/' + p.name.split(':')[0]
+        self._meta_placeholders[p] = tf.placeholder(p.dtype, p.shape, name)
+
+      # Create the loss function for meta-optimization.
+
+      updated_loss = self._loss
+      updated_variables = learner.variables
+      for i in range(optimization_steps):
+        gradients = tf.gradients(updated_loss, updated_variables)
+        updated_variables = [
+            v if g is None else v - self._learning_rate * g
+            for v, g in zip(updated_variables, gradients)
+        ]
+        replacements = dict(
+            (tf.convert_to_tensor(v1), v2)
+            for v1, v2 in zip(learner.variables, updated_variables))
+        if i == optimization_steps - 1:
+          # In the final loss, use different placeholders for all inputs so the loss will be
+          # computed from a different batch.
+
+          for p in self._meta_placeholders:
+            replacements[p] = self._meta_placeholders[p]
+        updated_loss = tf.contrib.graph_editor.graph_replace(
+            self._loss, replacements)
+      self._meta_loss = updated_loss
+
+      # Create the optimizer for meta-optimization.
+
+      self._global_step = tf.placeholder(tf.int32, [])
+      self._tf_optimizer = optimizer._create_optimizer(self._global_step)
+      self._train_op = self._tf_optimizer.minimize(self._meta_loss)
+      self._session = tf.Session()
+
+  def __del__(self):
+    if '_model_dir_is_temp' in dir(self) and self._model_dir_is_temp:
+      shutil.rmtree(self.model_dir)
+
+  def fit(self,
+          steps,
+          max_checkpoints_to_keep=5,
+          checkpoint_interval=600,
+          restore=False):
+    """Perform meta-learning to train the model.
+
+    Parameters
+    ----------
+    steps: int
+      the number of steps of meta-learning to perform
+    max_checkpoints_to_keep: int
+      the maximum number of checkpoint files to keep.  When this number is reached, older
+      files are deleted.
+    checkpoint_interval: float
+      the time interval at which to save checkpoints, measured in seconds
+    restore: bool
+      if True, restore the model from the most recent checkpoint and continue training
+      from there.  If False, retrain the model from scratch.
+    """
+    with self._graph.as_default():
+      self._session.run(tf.global_variables_initializer())
+      if restore:
+        self.restore()
+      saver = tf.train.Saver(
+          self.learner.variables, max_to_keep=max_checkpoints_to_keep)
+      checkpoint_index = 0
+      checkpoint_time = time.time()
+
+      # Main optimization loop.
+
+      for i in range(steps):
+        self.learner.select_task()
+        feed_dict = self.learner.get_batch()
```

Comment:
I'd interpret it as saying that I currently have a hardcoded batch size of 1.  Yes, that secondary level of batching still needs to be implemented.